### PR TITLE
Add Yaml::PARSE_EXCEPTION_ON_DUPLICATE to throw exceptions on duplicates

### PR DESCRIPTION
**Pull request by alexpott**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | https://github.com/symfony/symfony/issues/19526 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

**Comment by iltar** [comment]
I'd love this feature. In a few big applications with 25+ developers, mistakes like this happen every once in a while. I love to see them crash instead of silently continuing.

**Comment by Ener-Getick** [comment]
As proposed by other people, i would prefer deprecate this and throw an exception in symfony 4.0.

**Comment by nicolas-grekas** [comment]
@alexpott OK to replace this flag by a deprecation?

File: src/Symfony/Component/Yaml/Inline.php [hunk 1]
```diff
@@ -475,6 +475,8 @@ private static function parseMapping($mapping, $flags, &$i = 0, $references = ar
                         // are processed sequentially.
                         if (!isset($output[$key])) {
                             $output[$key] = $value;
+                        } else {
+                            @trigger_error(sprintf('Duplicate key "%s" detected whilst parsing YAML. Silent handling of duplicates in YAML is deprecated since version 3.3 and will cause an exception in 4.0.', $key), E_USER_DEPRECATED);
```
Comment (ogizanagi) [inline src/Symfony/Component/Yaml/Inline.php]: Should be `since 3.2`, right ?

Comment (ogizanagi) [inline src/Symfony/Component/Yaml/Inline.php]: We should probably also indicates the thrown exception will be a `Symfony\Component\Yaml\Exception\ParseException`

**Comment by alexpott** [comment]
The failure in fabbot.io is on a line not changed by this patch - I guess src/Symfony/Component/Yaml/Tests/Fixtures/YtsSpecificationExamples.yml does not comply with the standard.

**Comment by fabpot** [comment]
👍

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 2]
```diff
@@ -771,7 +771,7 @@ public function testScalarInSequence()
      * @see http://yaml.org/spec/1.2/spec.html#id2759572
      * @see http://yaml.org/spec/1.1/#id932806
      */
-    public function testMappingDuplicateKeyBlock()
+    public function testLegacyMappingDuplicateKeyBlock()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: missing `@group legacy` annotation (and no need to rename the test case then)
same below

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 3]
```diff
@@ -803,6 +803,84 @@ public function testMappingDuplicateKeyFlow()
         $this->assertSame($expected, Yaml::parse($input));
     }
 
+    /**
+     * @dataProvider getParseExceptionOnDuplicateData
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: missing note saying that this should turned to an `@expectedException` assertion on 4.0, isn't it?

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 4]
```diff
@@ -803,6 +803,84 @@ public function testMappingDuplicateKeyFlow()
...
+     */
+    public function testParseExceptionOnDuplicate($input, $duplicate_key)
+    {
+        $deprecations = array();
+        set_error_handler(function ($type, $msg) use (&$deprecations) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: use `ErrorAssert::assertDeprecationsAreTriggered` instead (grep the code to find usages)

File: src/Symfony/Component/Yaml/Tests/ParserTest.php (hunk 3 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: Not sure we need it, the test will fail anyway when the deprecation will be removed.

**Comment by xabbuh** [comment]
Please also update the upgrade files for 3.2 and 4.0.

File: UPGRADE-4.0.md [hunk 5]
```diff
@@ -230,6 +230,10 @@ Yaml
  * The `!!php/object` tag to indicate dumped PHP objects was removed in favor of
    the `!php/object` tag.
 
+ * Support for silently ignoring duplicate keys in YAML has been deprecated and
```
Comment (xabbuh) [inline UPGRADE-4.0.md]: In 4.0 it will be longer deprecated.

File: src/Symfony/Component/Yaml/Tests/ParserTest.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: We don't *need* it, but it may help the one doing the cleanup when preparing 4.0...

**Comment by fabpot** [comment]
Thank you @alexpott.

**Comment by stof** [comment]
IIRC, the Yaml spec explicitly says that parsers should use the first key when finding duplicates (and we made a bugfix in the past to respect it instead of using the last one). So this deprecation is a violation of the spec

**Comment by Ener-Getick** [comment]
@stof from what i see, the spec says that it should be considered as an error:
> If both notations are used as keys in the same mapping, only a YAML processor which recognizes integer formats would correctly flag the duplicate key as an error.

> Technically, YAML therefore complies with the JSON spec, choosing to treat duplicates as an error. In practice, since JSON is silent on the semantics of such duplicates, the only portable JSON files are those with unique keys, which are therefore valid YAML files.

**Comment by ivangretsky** [comment]
Good day!
Was looking for exact same feature. It is said to be in the master, but I cannot [find it](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Yaml/Yaml.php#L23). Am I missing something?

**Comment by xabbuh** [comment]
@ivangretsky There is no flag controlling this. The parser just triggers a deprecation when detecting duplicated keys on 3.2 or higher and will throw an exception starting with Symfony 4.0.

**Pull request closed without merging**